pkg/metrics/query.go

```diff
@@ -0,0 +1,188 @@
+package metrics
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	kapierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/prometheus/common/model"
+)
+
+const (
+	serviceCAinjectionDataKey             = "service-ca.crt"
+	serviceCAinjectCABundleAnnotationName = "service.beta.openshift.io/inject-cabundle"
+	metricsBaseURL                        = "https://prometheus-k8s.openshift-monitoring.svc:9091"
+)
+
+type PrometheusClient struct {
+	Client      *http.Client
+	URL         string
+	BearerToken string
+}
+
+func NewPrometheusClient(caBundle []byte, query, bearerToken string) (*PrometheusClient, error) {
+	url := fmt.Sprintf("%s/api/v1/query?%s", metricsBaseURL, (url.Values{"query": []string{query}}).Encode())
+	certPool := x509.NewCertPool()
+	certPool.AppendCertsFromPEM(caBundle)
+	tlsConfig := &tls.Config{
+		RootCAs: certPool,
+	}
+
+	tr := &http.Transport{
+		TLSClientConfig: tlsConfig,
+	}
+	tlsClient := &http.Client{Transport: tr}
+	return &PrometheusClient{
+		Client:      tlsClient,
+		URL:         url,
+		BearerToken: bearerToken,
+	}, nil
+}
+
+func LocatePrometheus(client *kubernetes.Clientset) (string, bool) {
+	_, err := client.CoreV1().Services("openshift-monitoring").Get("prometheus-k8s", metav1.GetOptions{})
+	if kapierrs.IsNotFound(err) {
+		return "", false
+	}
+
+	var bearerToken string
+	err = wait.Poll(time.Second*1, time.Second*30, func() (bool, error) {
+		secrets, err := client.CoreV1().Secrets("openshift-monitoring").List(metav1.ListOptions{})
+		if err != nil {
+			return false, fmt.Errorf("could not list secrets in openshift-monitoring namespace")
```

Comment:
Please embed the original error in the message to simplify failure analysis.

Reply:
updated